server/src/main/java/de/uftos/repositories/ucdl/parser/DefinitionParser.java

```diff
@@ -0,0 +1,691 @@
+                  default -> throw new IllegalStateException();
+                };
+              case "EQUATION":
+                params = new ArrayList<>();
+                Node firstElement = root.jjtGetChild(0);
+                Node secondElement = root.jjtGetChild(1).jjtGetChild(1);
+                String comparator =
+                    ((SimpleNode) root.jjtGetChild(1).jjtGetChild(0)).jjtGetFirstToken().image;
+
+                params.add(buildAst(firstElement, parameters));
+                params.add(buildAst(secondElement, parameters));
+                UcdlToken token = getComparatorToken(comparator, params);
+                return new OperatorDto(token, params);
+              default:
+                throw new IllegalStateException();
+            }
+          default:
+            throw new IllegalStateException();
+
+        }
+      }
+      //includes cases: "CONTROLSEQUENCELIST", "CONTROLSEQUENCERETURN",
+      // "OPTIONALIMPLIES", "OPTIONALOR", "OPTIONALAND",
+      // "ELEMENTINSET", "EQUATION", "ELEMENTEQUATION", "ELEMENTATTRIBUTELIST",
+      // "SETNAME", "SETMODIFICATION", "NUMBERLIST", "ATTRIBUTE", "FILTERLIST"
+      default -> throw new IllegalStateException();
+    }
+  }
+
+  //checks that all returnValues are the same and returns the value
+  // (throws an exception if true and false are returnValues inside the body)
+  private static boolean getReturnValue(List<AbstractSyntaxTreeDto> body) throws ParseException {
+    boolean and = true;
+    boolean or = false;
+    for (AbstractSyntaxTreeDto ast : body) {
+      if (ast.getToken() == UcdlToken.FOR || ast.getToken() == UcdlToken.IF) {
+        ControlSequenceDto cs = (ControlSequenceDto) ast;
+        and = and && cs.returnValue();
+        or = or || cs.returnValue();
+      }
+    }
+    if (and != or) {
+      throw new ParseException(
+          "All control sequences in the body of a control sequence need to return"
+              + " the same boolean value!");
+    }
+    return and;
+  }
+
+  private static UcdlToken getComparatorToken(String comparator, List<AbstractSyntaxTreeDto> params)
+      throws ParseException {
+    UcdlToken token = switch (comparator) {
+      case ">" -> UcdlToken.GREATER;
+      case "<" -> UcdlToken.SMALLER;
+      case ">=" -> UcdlToken.GREATER_EQUALS;
+      case "<=" -> UcdlToken.SMALLER_EQUALS;
+      case "=", "==" -> UcdlToken.EQUALS;
+      case "!=" -> UcdlToken.NOT_EQUALS;
+      default -> throw new IllegalStateException();
+    };
+    if (token == UcdlToken.GREATER || token == UcdlToken.SMALLER
+        || token == UcdlToken.GREATER_EQUALS || token == UcdlToken.SMALLER_EQUALS) {
+      for (AbstractSyntaxTreeDto ast : params) {
+
+        if (ast.getToken() != UcdlToken.NUMBER && ast.getToken() != UcdlToken.SIZE
+            && !(ast.getToken() == UcdlToken.ELEMENT
+            && ((ElementDto) ast).type() == ResourceType.NUMBER)) {
```

Comment:
Shouldn't this be "if the type is *not* `ResourceType.NUMBER`"?

Reply:
It is already negated: `!(ast.getToken() == UcdlToken.ELEMENT && ((ElementDto) ast).type() == ResourceType.NUMBER)`

Reply:
Oh, I overlooked that that whole expression is in parenthesis. This is probably another place where you should refactor the condition :P